File: vbnet/perfect-numbers/PerfectNumbers.vb
Imports System
Imports System.Linq

Public Enum Classification
    Perfect
    Abundant
    Deficient
End Enum

Public Module PerfectNumbers
    Public Function Classify(number As Integer) As Classification
        If number < 1 
            Throw New ArgumentOutOfRangeException()
        End If
        
        Dim sum As Integer = AliquotSum(number)
        If sum > number 
            Return Classification.Abundant
        ElseIf sum < number
            Return Classification.Deficient
        Else
            Return Classification.Perfect
        End If
    End Function
    
    Private Function AliquotSum(number As Integer) As Integer
        Return Enumerable.Range(1, number / 2).Where(Function(x) number Mod x = 0).Sum()
    End Function
End Module

File: prolog/perfect-numbers/perfect_numbers.pl
:- use_module(library(clpfd)).

factors(1, []) :- !.
factors(Number, [1|Factors]) :-
    MaxSmallFactor is floor(sqrt(Number)),
    findall(SmallFactor, (between(2, MaxSmallFactor, SmallFactor), Number mod SmallFactor =:= 0), SmallFactors),
    findall(BigFactor, (member(Factor, SmallFactors), BigFactor is Number // Factor, BigFactor < Number, BigFactor =\= Factor), BigFactors),
    append(SmallFactors, BigFactors, FactorsUnsorted),
    sort(FactorsUnsorted, Factors).

aliquot_sum(Number, AliquotSum) :-    
    factors(Number, Factors),
    sum_list(Factors, AliquotSum).

classify(Number, AliquotSum, abundant) :- AliquotSum #> Number, !.
classify(Number, AliquotSum, deficient) :- AliquotSum #< Number, !.
classify(_, _, perfect) :- !.

classify(Number, Classification) :-
    Number #> 0,
    aliquot_sum(Number, AliquotSum),
    classify(Number, AliquotSum, Classification).
